analysis/subset_data/SubsetBam.wdl

```diff
@@ -0,0 +1,169 @@
+workflow SubsetBam {
+  # Bam to subset
+  File aligned_gold_bam
+
+  # How many cells to grab from the largest and smallest end
+  Int num_large_cells
+  Int num_small_cells
+
+  # How many reads a cell must have before it can be selected
+  Int? min_reads_per_cell
+
+  # Array of comma separated tag values to put in each fastq e.g ["ORIG", "SR", "CB,UB"] "ORIG" means to
+  # grab the original sequence/qualities from the read
+  Array[String] barcode_tags_to_split_by
+  Array[String] quality_tags_to_split_by
+  Array[String]? separators_to_split_by
+
+  Array[String]? final_fastq_basenames
+
+  Int bam_size = ceil(size(aligned_gold_bam, "GB"))
+  Int additional_disk = 10
+
+  # Figure out which cell barcodes to select from bam
+  call SelectCells {
+    input:
+      aligned_bam = aligned_gold_bam,
+      num_large_cells = num_large_cells,
+      num_small_cells = num_small_cells,
+      min_reads_per_cell = min_reads_per_cell,
+      disk = bam_size + additional_disk
+  }
+
+  # Grab reads from selected cell barcodes
+  call SubsetBamByTag {
+    input:
+      aligned_bam = aligned_gold_bam,
+      tag_values = SelectCells.tag_values,
+      disk = (bam_size * 2) + additional_disk # 2 * full bam size here as that would be the max subset size...
+  }
+
+  Int subset_bam_size = ceil(size(SubsetBamByTag.subset_bam, "GB"))
+  # Split bam into bams described by barcode_tags_to_split_by, quality_tags_to_split_by, separators_to_split_by
+  call SplitBam {
+    input:
+    subset_bam = SubsetBamByTag.subset_bam,
+    barcode_tags_to_split_by = barcode_tags_to_split_by,
+    quality_tags_to_split_by = quality_tags_to_split_by,
+    separators_to_split_by = separators_to_split_by,
+    disk = (subset_bam_size * 2) + additional_disk # max split size would be 2 * subset size
+  }
+
+  # Revert bams into fastqs
+  scatter (index in range(length(SplitBam.split_bams))) {
+    String base = if (defined(final_fastq_basenames)) then select_first([final_fastq_basenames])[index] else basename(SplitBam.split_bams[index])
+    Int split_bam_size = ceil(size(SplitBam.split_bams[index], "GB"))
+
+    call SamToFastq {
+      input:
+        bam = SplitBam.split_bams[index],
+        output_basename = base,
+        disk = (split_bam_size * 2) + additional_disk  # fast q shouldn't be larger than the bam size
+    }
+  }
+}
+
+task SelectCells {
+  File aligned_bam
+  Int num_large_cells
+  Int num_small_cells
+  Int? min_reads_per_cell
```

Comment:
Ah ok; now I see what `min_reads_per_cell` was doing; I think you should define it here as 100 instead of in the bash line. 

I also think we should:
(1) throw an error if tags_to_filter ends up with fewer than the expected number of tags
(2) only apply the minimum to the "large" cell barcodes. 
(2) make sure we're not taking overlapping sets i.e. if the histogram contains 150 cell barcodes, we might take 0:100 and 50:150, with 50 represented twice.

Reply:
We should about why I provided the default value the way that I did because I've had this argument with multiple people and would like your opinion on it as well.

(1) 👍  I already have that in a commit
(2) see earlier comment about possibly having a min for "large" and "small" cell barcodes
(2/3*) The error check i write for (1) should cover this as well

Reply:
(1) great
(2) Yep, you clarified. 
(3) This is the case if you look at the size of the unique set of tags, but otherwise the length alone wouldn't necessarily capture this, due to the possibility of redundant tags. So I mentioned it separately. Glad that you've already covered this :+1: